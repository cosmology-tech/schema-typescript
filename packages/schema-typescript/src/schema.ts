import generate from "@babel/generator";
import * as t from "@babel/types";

import { SchemaTSContext, type SchemaTSOptions } from "./context";
import type { JSONSchema } from "./types";
import { getTypeNameSafe, isValidIdentifier, isValidIdentifierCamelized, makeCommentLine, shouldInclude, toCamelCase } from "./utils";

const getSchemaTypeNameSafe = (ctx: SchemaTSContext, str: string): string => {
  return getTypeNameSafe(ctx.options.namingStrategy, str);
};

const identifier = (name: string, typeAnnotation: t.TSTypeAnnotation) => {
  const i = t.identifier(name);
  i.typeAnnotation = typeAnnotation;
  return i;
};

const anyOrObjectWithUnknownProps = (ctx: SchemaTSContext) => {
  return ctx.options.strictTypeSafety ? t.tsTypeLiteral([
    t.tsIndexSignature(
      [
        identifier('key', t.tsTypeAnnotation(
          t.tsStringKeyword()
        ))
      ],
      t.tsTypeAnnotation(t.tsUnknownKeyword())
    )
  ]) : t.tsAnyKeyword();
}

export function generateTypeScriptTypes(schema: JSONSchema, options?: Partial<SchemaTSOptions>): t.ExportNamedDeclaration[] {
  const ctx = new SchemaTSContext(options, schema, schema, []);
  return generateInterfaces(ctx, schema);
}

export function generateTypeScript(schema: JSONSchema, options?: Partial<SchemaTSOptions>): string {
  const ctx = new SchemaTSContext(options, schema, schema, []);
  const interfaces = generateInterfaces(ctx, schema);
  return generate(t.file(t.program(interfaces))).code;
}

export function generateInterfaces(ctx: SchemaTSContext, schema: JSONSchema): t.ExportNamedDeclaration[] {
  const interfaces = [];

  try {
    // Process both $defs and definitions
    const definitions = schema.$defs || schema.definitions || {};
    for (const key in definitions) {
      if (shouldInclude(key, {
        include: ctx.options.include,
        exclude: ctx.options.exclude,
      })) {
        interfaces.push(createInterfaceDeclaration(ctx, key, definitions[key]));
      }
    }
  } catch (e) {
    console.error('Error processing interfaces');
    throw e;
  }

  // Process the main schema
  const title = schema.title;
  if (!title) {
    throw new Error('schema or options require a title');
  }
  if (shouldInclude(title, {
    include: ctx.options.include,
    exclude: ctx.options.exclude,
  })) {
    interfaces.push(createInterfaceDeclaration(ctx, title, schema));
  }
  return interfaces;
}

const createExportDeclarationForType = (
  ctx: SchemaTSContext,
  name: string,
  schema: JSONSchema,
  node: t.Declaration,
) => {
  const result =  t.exportNamedDeclaration(node);

  if (ctx.options.includeTypeComments && schema.description) {

    if (name.includes('.')) {
      // for complex names, let's add them for clarity/mapping
      result.leadingComments = [makeCommentLine(name)[0], makeCommentLine(schema.description)[0]]
    } else {
      result.leadingComments = makeCommentLine(schema.description)
    }

  }
  return result;
}

function createInterfaceDeclaration(
  ctx: SchemaTSContext,
  name: string,
  schema: JSONSchema
): t.ExportNamedDeclaration {
  // Handle standard properties if they exist
  let bodyElements: any = [];
  if (schema.properties) {
    const properties = schema.properties;
    const required = schema.required || [];
    bodyElements = Object.keys(properties).map(key => {
      const prop = properties[key];
      return createPropertySignature(ctx, key, prop, required, schema);
    });
  }

  // Handling additionalProperties if they exist
  if (schema.additionalProperties) {
    const additionalType = typeof schema.additionalProperties === 'boolean' ?
      t.tsStringKeyword() : getTypeForProp(ctx, schema.additionalProperties, [], schema);
    const indexSignature = t.tsIndexSignature(
      [t.identifier("key")], // index name, can be any valid name
      t.tsTypeAnnotation(additionalType)
    );
    indexSignature.parameters[0].typeAnnotation = t.tsTypeAnnotation(t.tsStringKeyword());
    bodyElements.push(indexSignature);
  }

  // Handling oneOf, anyOf, allOf if properties are not defined
  if (!schema.properties && (schema.oneOf || schema.anyOf || schema.allOf)) {
    const types = [];
    if (schema.oneOf) {
      types.push(getTypeForProp(ctx, { oneOf: schema.oneOf }, [], schema));
    }
    if (schema.anyOf) {
      types.push(getTypeForProp(ctx, { anyOf: schema.anyOf }, [], schema));
    }
    if (schema.allOf) {
      types.push(getTypeForProp(ctx, { allOf: schema.allOf }, [], schema));
    }

    // Create a union type if multiple types are generated
    const combinedType = types.length > 1 ? t.tsUnionType(types) : types[0];

    // Create a type alias instead of an interface if we're only handling these constructs
    const typeAlias = t.tsTypeAliasDeclaration(t.identifier(getSchemaTypeNameSafe(ctx, name)), null, combinedType);
    return createExportDeclarationForType(ctx, name, schema, typeAlias);
  }

  // Finally, create the interface declaration if there are any body elements
  if (bodyElements.length > 0) {
    const interfaceDeclaration = t.tsInterfaceDeclaration(
      t.identifier(getSchemaTypeNameSafe(ctx, name)),
      null,
      [],
      t.tsInterfaceBody(bodyElements)
    );
    return createExportDeclarationForType(ctx, name, schema, interfaceDeclaration);
  }

  if (schema.type) {
    return createExportDeclarationForType(ctx, name, schema, t.tsTypeAliasDeclaration(t.identifier(getSchemaTypeNameSafe(ctx, name)), null, getTypeForProp(ctx, schema, [], schema)))    
  }

  if (ctx.options.overrides && Object.prototype.hasOwnProperty.call(ctx.options.overrides, name)) {
    return createExportDeclarationForType(ctx, name, schema,
      t.tsTypeAliasDeclaration(t.identifier(getSchemaTypeNameSafe(ctx, name)), null,
      getTypeForProp(ctx, ctx.options.overrides[name], [], schema)
    ));
  }

  // Fallback to exporting a basic type if nothing else is possible
  // console.warn(`No properties or type definitions found for ${name}, defaulting to 'any'.`);
  return createExportDeclarationForType(ctx, name, schema, t.tsTypeAliasDeclaration(t.identifier(getSchemaTypeNameSafe(ctx, name)), null, t.tsAnyKeyword()))
}

function createPropertySignature(
  ctx: SchemaTSContext,
  key: string,
  prop: JSONSchema,
  required: string[],
  schema: JSONSchema
): t.TSPropertySignature {
  let camelCaseFn: (str: string) => string = toCamelCase;
  if (ctx.options.camelCaseFn) camelCaseFn = ctx.options.camelCaseFn;
  const name = ctx.options.camelCase ? camelCaseFn(key) : key;
  const propType = getTypeForProp(ctx, prop, required, schema);
  let identifier: t.Identifier | t.StringLiteral;
  let isIdent: boolean;
  if (ctx.options.camelCase) {
    isIdent = isValidIdentifierCamelized(key);
  } else {
    isIdent = isValidIdentifier(key);
  }
  identifier = isIdent ? t.identifier(name) : t.stringLiteral(key);
  const propSig = t.tsPropertySignature(
    identifier,
    t.tsTypeAnnotation(propType)
  );
  propSig.optional = !required.includes(key);
  if (ctx.options.includePropertyComments && prop.description) {
    propSig.leadingComments = makeCommentLine(prop.description);
  }

  return propSig;
}

function getTypeForProp(ctx: SchemaTSContext, prop: JSONSchema, required: string[], schema: JSONSchema): t.TSType {
  if (prop.$ref) {
    return resolveRefType(ctx, prop.$ref, schema);
  }

  if (prop.enum) {
    const enumType = prop.enum.map(enumValue => t.tsLiteralType(t.stringLiteral(enumValue)));
    return t.tsUnionType(enumType);
  }

  if (prop.const) {
    return t.tsLiteralType(t.stringLiteral(prop.const));
  }

  if (prop.type) {
    if (Array.isArray(prop.type)) {
      const arrayType = prop.type.map(type => getTypeForProp(ctx, { type, items: prop.items }, [], schema));
      return t.tsUnionType(arrayType);
    }

    switch (prop.type) {
      case 'string':
        return t.tsStringKeyword();
      case 'number':
      case 'integer':
        return t.tsNumberKeyword();
      case 'boolean':
        return t.tsBooleanKeyword();
      case 'null':
        return t.tsNullKeyword();
      case 'array':
        if (prop.items) {
          return t.tsArrayType(getTypeForProp(ctx, prop.items, required, schema));
        } else {
          throw new Error('Array items specification is missing');
        }
      case 'object':
        if (prop.properties) {
          const nestedProperties = prop.properties;
          const nestedRequired = prop.required || [];
          const typeElements = Object.keys(nestedProperties).map(nestedKey => {
            const nestedProp = nestedProperties[nestedKey];
            return createPropertySignature(ctx, nestedKey, nestedProp, nestedRequired, schema);
          });
          return t.tsTypeLiteral(typeElements);
        } else {
          // Handle dynamic properties with strict type safety option
          return anyOrObjectWithUnknownProps(ctx);
        }
      default:
        return t.tsAnyKeyword();
    }
  }

  // Handling oneOf, anyOf, allOf
  if (prop.anyOf) {
    const types = prop.anyOf.map((subProp) => getTypeForProp(ctx, subProp, required, schema));
    return t.tsUnionType(types);
  }
  if (prop.allOf) {
    const types = prop.allOf.map((subProp) => getTypeForProp(ctx, subProp, required, schema));
    return t.tsIntersectionType(types);
  }
  if (prop.oneOf) {
    const types = prop.oneOf.map((subProp) => getTypeForProp(ctx, subProp, required, schema));
    return t.tsUnionType(types);
  }

  // Fallback when no types are defined
  return t.tsAnyKeyword()

}

function getTypeReferenceFromSchema(ctx: SchemaTSContext, schema: JSONSchema, definitionName: string): t.TSType | null {
  if (definitionName) {

    // if (ctx.options.overrides && Object.prototype.hasOwnProperty.call(ctx.options.overrides, definitionName)) {
    //   return getTypeForProp(ctx, ctx.options.overrides[definitionName], [], schema);
    // }

    if (schema.$defs && schema.$defs[definitionName]) {
      return t.tsTypeReference(t.identifier(getSchemaTypeNameSafe(ctx, definitionName)));
    } else if (schema.definitions && schema.definitions[definitionName]) {
      return t.tsTypeReference(t.identifier(getSchemaTypeNameSafe(ctx, definitionName)));
    }
  }
  return null;  // Return null if no type reference is found
}


function resolveRefType(ctx: SchemaTSContext, ref: string, schema: JSONSchema): t.TSType {
  const path = ref.split('/');
  const definitionName = path.pop();

  // Try to resolve the type reference from the local schema
  const localTypeReference = getTypeReferenceFromSchema(ctx, schema, definitionName);
  if (localTypeReference) {
    return localTypeReference;
  }

  // Try to resolve the type reference from the root schema
  const rootTypeReference = getTypeReferenceFromSchema(ctx, ctx.root, definitionName);
  if (rootTypeReference) {
    return rootTypeReference;
  }

  // If no definitions are found in either schema, throw an error
  throw new Error(`Reference ${ref} not found in definitions or $defs.`);
}
