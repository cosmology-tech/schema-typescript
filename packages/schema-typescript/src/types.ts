export interface JSONSchema {
  $schema?: string;
  $ref?: string;
  title?: string;
  properties?: { [key: string]: JSONSchema };
  required?: string[];
  type?: string;
  const?: string;
  enum?: string[];
  items?: JSONSchema;
  $defs?: { [key: string]: JSONSchema };  // (JSON Schema Draft 2019-09 and later)
  definitions?: { [key: string]: JSONSchema }; // (JSON Schema Draft-04 to Draft-07)
  additionalProperties?: boolean | JSONSchema;
  anyOf?: JSONSchema[];
  allOf?: JSONSchema[];
  oneOf?: JSONSchema[];
  description?: string;
  default?: any;
  format?: string;
}

export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Array<infer U>
      ? Array<DeepPartial<U>> // If it's an array, make each element a DeepPartial
      : T[P] extends ReadonlyArray<infer U>
      ? ReadonlyArray<DeepPartial<U>> // Handle readonly arrays
      : T[P] extends object
      ? DeepPartial<T[P]> // Apply DeepPartial recursively if it's an object
      : T[P]; // Otherwise, just make it optional
};
