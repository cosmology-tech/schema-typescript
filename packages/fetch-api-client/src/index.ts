import fetch from 'isomorphic-fetch';

interface RequestOptions<Params> {
  hostname: string;
  path: string;
  headers?: { [key: string]: string };
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
  params?: Params;
  timeout?: number;
}

export interface APIClientOptions {
  restEndpoint: string;
}

export class APIClient {
  private baseUrl: string;
  private defaultTimeout: number = 10000; // 10 seconds

  constructor(options: APIClientOptions) {
    this.baseUrl = options.restEndpoint;
  }

  private buildUrl(endpoint: string, query?: { [key: string]: any }): string {
    const url = new URL(endpoint, this.baseUrl);
    if (query) {
      Object.keys(query).forEach(key => url.searchParams.append(key, query[key]));
    }
    return url.toString();
  }

  private async request<Resp>(options: RequestOptions<any>): Promise<Resp> {
    const { path, headers, method, params, timeout } = options;
    const url = this.buildUrl(path, method === 'GET' && params);
    const fetchOptions: RequestInit = {
      method,
      headers,
      body: method !== 'GET' && method !== 'DELETE' ? JSON.stringify(params) : null,
    };

    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), timeout || this.defaultTimeout);
    fetchOptions.signal = controller.signal;

    const response = await fetch(url, fetchOptions);
    clearTimeout(id);

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    return response.json() as Promise<Resp>;
  }

  get<Resp = unknown>(endpoint: string, query?: { [key: string]: any }, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'GET',
      headers: opts.headers,
      timeout: opts.timeout || this.defaultTimeout,
      params: query, // Correctly use query parameters for GET
      hostname: '',
    });
  }

  post<Resp = unknown, Params = any>(endpoint: string, params: Params, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...opts.headers,
      },
      params,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }

  put<Resp = unknown, Params = any>(endpoint: string, params: Params, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        ...opts.headers,
      },
      params,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }

  patch<Resp = unknown, Params = any>(endpoint: string, params: Params, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        ...opts.headers,
      },
      params,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }

  delete<Resp = unknown>(endpoint: string, opts: { headers?: { [key: string]: string }, timeout?: number } = {}): Promise<Resp> {
    return this.request<Resp>({
      path: endpoint,
      method: 'DELETE',
      headers: opts.headers,
      timeout: opts.timeout || this.defaultTimeout,
      hostname: '',
    });
  }
}
