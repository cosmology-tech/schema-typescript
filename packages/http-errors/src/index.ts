export class HTTPError extends Error {
  constructor(message: string, public status: number) {
    super(`${message} | Status: ${status}`);
    this.name = this.constructor.name;
  }
}
    
export class BadRequestError extends HTTPError {
  constructor() { super('Bad Request', 400); }
}
export class UnauthorizedError extends HTTPError {
  constructor() { super('Unauthorized', 401); }
}
export class PaymentRequiredError extends HTTPError {
  constructor() { super('Payment Required', 402); }
}
export class ForbiddenError extends HTTPError {
  constructor() { super('Forbidden', 403); }
}
export class NotFoundError extends HTTPError {
  constructor() { super('Not Found', 404); }
}
export class MethodNotAllowedError extends HTTPError {
  constructor() { super('Method Not Allowed', 405); }
}
export class NotAcceptableError extends HTTPError {
  constructor() { super('Not Acceptable', 406); }
}
export class ProxyAuthenticationRequiredError extends HTTPError {
  constructor() { super('Proxy Authentication Required', 407); }
}
export class RequestTimeoutError extends HTTPError {
  constructor() { super('Request Timeout', 408); }
}
export class ConflictError extends HTTPError {
  constructor() { super('Conflict', 409); }
}
export class GoneError extends HTTPError {
  constructor() { super('Gone', 410); }
}
export class LengthRequiredError extends HTTPError {
  constructor() { super('Length Required', 411); }
}
export class PreconditionFailedError extends HTTPError {
  constructor() { super('Precondition Failed', 412); }
}
export class PayloadTooLargeError extends HTTPError {
  constructor() { super('Payload Too Large', 413); }
}
export class URITooLongError extends HTTPError {
  constructor() { super('URI Too Long', 414); }
}
export class UnsupportedMediaTypeError extends HTTPError {
  constructor() { super('Unsupported Media Type', 415); }
}
export class RangeNotSatisfiableError extends HTTPError {
  constructor() { super('Range Not Satisfiable', 416); }
}
export class ExpectationFailedError extends HTTPError {
  constructor() { super('Expectation Failed', 417); }
}
export class ImATeapotError extends HTTPError {
  constructor() { super("I'm a Teapot", 418); }
}
export class MisdirectedRequestError extends HTTPError {
  constructor() { super('Misdirected Request', 421); }
}
export class UnprocessableEntityError extends HTTPError {
  constructor() { super('Unprocessable Entity', 422); }
}
export class LockedError extends HTTPError {
  constructor() { super('Locked', 423); }
}
export class FailedDependencyError extends HTTPError {
  constructor() { super('Failed Dependency', 424); }
}
export class UpgradeRequiredError extends HTTPError {
  constructor() { super('Upgrade Required', 426); }
}
export class PreconditionRequiredError extends HTTPError {
  constructor() { super('Precondition Required', 428); }
}
export class TooManyRequestsError extends HTTPError {
  constructor() { super('Too Many Requests', 429); }
}
export class RequestHeaderFieldsTooLargeError extends HTTPError {
  constructor() { super('Request Header Fields Too Large', 431); }
}
export class UnavailableForLegalReasonsError extends HTTPError {
  constructor() { super('Unavailable For Legal Reasons', 451); }
}
export class InternalServerError extends HTTPError {
  constructor() { super('Internal Server Error', 500); }
}
export class NotImplementedError extends HTTPError {
  constructor() { super('Not Implemented', 501); }
}
export class BadGatewayError extends HTTPError {
  constructor() { super('Bad Gateway', 502); }
}
export class ServiceUnavailableError extends HTTPError {
  constructor() { super('Service Unavailable', 503); }
}
export class GatewayTimeoutError extends HTTPError {
  constructor() { super('Gateway Timeout', 504); }
}
export class HTTPVersionNotSupportedError extends HTTPError {
  constructor() { super('HTTP Version Not Supported', 505); }
}
export class VariantAlsoNegotiatesError extends HTTPError {
  constructor() { super('Variant Also Negotiates', 506); }
}
export class InsufficientStorageError extends HTTPError {
  constructor() { super('Insufficient Storage', 507); }
}
export class LoopDetectedError extends HTTPError {
  constructor() { super('Loop Detected', 508); }
}
export class NotExtendedError extends HTTPError {
  constructor() { super('Not Extended', 510); }
}
export class NetworkAuthenticationRequiredError extends HTTPError {
  constructor() { super('Network Authentication Required', 511); }
}
    
  
export function createHTTPError(status: number): HTTPError {
  switch (status) {
  case 400: return new BadRequestError();
  case 401: return new UnauthorizedError();
  case 402: return new PaymentRequiredError();
  case 403: return new ForbiddenError();
  case 404: return new NotFoundError();
  case 405: return new MethodNotAllowedError();
  case 406: return new NotAcceptableError();
  case 407: return new ProxyAuthenticationRequiredError();
  case 408: return new RequestTimeoutError();
  case 409: return new ConflictError();
  case 410: return new GoneError();
  case 411: return new LengthRequiredError();
  case 412: return new PreconditionFailedError();
  case 413: return new PayloadTooLargeError();
  case 414: return new URITooLongError();
  case 415: return new UnsupportedMediaTypeError();
  case 416: return new RangeNotSatisfiableError();
  case 417: return new ExpectationFailedError();
  case 418: return new ImATeapotError();
  case 421: return new MisdirectedRequestError();
  case 422: return new UnprocessableEntityError();
  case 423: return new LockedError();
  case 424: return new FailedDependencyError();
  case 426: return new UpgradeRequiredError();
  case 428: return new PreconditionRequiredError();
  case 429: return new TooManyRequestsError();
  case 431: return new RequestHeaderFieldsTooLargeError();
  case 451: return new UnavailableForLegalReasonsError();
  case 500: return new InternalServerError();
  case 501: return new NotImplementedError();
  case 502: return new BadGatewayError();
  case 503: return new ServiceUnavailableError();
  case 504: return new GatewayTimeoutError();
  case 505: return new HTTPVersionNotSupportedError();
  case 506: return new VariantAlsoNegotiatesError();
  case 507: return new InsufficientStorageError();
  case 508: return new LoopDetectedError();
  case 510: return new NotExtendedError();
  case 511: return new NetworkAuthenticationRequiredError();
  default: return new HTTPError('Unknown Error', status);
  }
}
  