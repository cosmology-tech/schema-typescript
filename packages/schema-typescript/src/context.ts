import deepmerge from 'deepmerge';

import type { DeepPartial, JSONSchema } from "./types";

export interface SchemaDefinitionOverrides {
  [key: string]: JSONSchema
}

export interface SchemaNamingStrategy {
  useLastSegment?: boolean;
  prefixStrip?: string | RegExp;
  renameMap?: { [originalName: string]: string };
}
export interface SchemaTSOptions {
  useSingleQuotes: boolean;
  camelCase?: boolean; // defaults to false
  camelCaseFn?: (str: string) => string; // optional function to convert keys to camelCase
  strictTypeSafety: boolean;  // true uses { [k: string]: unknown; }, false uses any
  overrides?: SchemaDefinitionOverrides;
  
  includeTypeComments?: boolean;
  includePropertyComments?: boolean;
  includeMethodComments?: boolean;

  namingStrategy?: SchemaNamingStrategy;

  // Include/Exclude types
  include?: string[];
  exclude?: string[];
}

export interface SchemaTSContextI {
  options: SchemaTSOptions;
  root: JSONSchema;
  schema: JSONSchema;
  parents: JSONSchema[]
}


export class SchemaTSContext implements SchemaTSContextI {
  options: SchemaTSOptions;
  root: JSONSchema;
  schema: JSONSchema;
  parents: JSONSchema[];

  constructor(
    options: Partial<SchemaTSOptions>,
    root: JSONSchema,
    schema: JSONSchema,
    parents: JSONSchema[] = []
  ) {
    this.options = getDefaultSchemaTSOptions(options);
    this.schema = schema;
    this.root = root;
    this.parents = parents;
  }

  // (currently not used)
  // Clone the context with the option to add a new parent
  clone(newParent: JSONSchema): SchemaTSContext {
    // Create a new array for parents to avoid mutation of the original array
    const newParents = [...this.parents, newParent];
    return new SchemaTSContext(this.options, this.root, this.schema, newParents);
  }
}

export const defaultSchemaTSOptions: SchemaTSOptions = { 
  useSingleQuotes: true, 
  camelCase: false,
  camelCaseFn: null,
  strictTypeSafety: true,
  exclude: [],
  include: [],
  includePropertyComments: false,
  includeMethodComments: false,
  includeTypeComments: false,
  namingStrategy: {
    useLastSegment: true
  }
};

export const getDefaultSchemaTSOptions = (options?: DeepPartial<SchemaTSOptions>): SchemaTSOptions => {
  return deepmerge(defaultSchemaTSOptions, options ?? {}) as SchemaTSOptions;
};
