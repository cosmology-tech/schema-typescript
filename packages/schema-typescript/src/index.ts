export * from './context';
export * from './schema';
export * from './types';
export * from './utils';
