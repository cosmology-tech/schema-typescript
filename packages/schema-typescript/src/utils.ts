import * as t from '@babel/types';
import { toPascalCase } from '@interweb-utils/casing';
import { minimatch } from 'minimatch';

import { SchemaNamingStrategy } from './context';

export const getTypeNameSafe = (strategy: SchemaNamingStrategy, str: string): string => {
  if (Object.prototype.hasOwnProperty.call(strategy.renameMap ?? {}, str)) {
    return toPascalCase(strategy.renameMap[str]);
  }

  if (strategy.useLastSegment) {
    const parts = str.split('.');
    if (parts.length > 1) {
      return toPascalCase(parts[parts.length - 1]);
    }
    return toPascalCase(str);
  }

  if (str.match(/\./)) return toPascalCase(str.replace(/\./g, '_'));

  return toPascalCase(str);
};

const globPattern = /\*+([^+@!?\*\[\(]*)/;

interface ShouldIncludeOptions {
  include: string[];
  exclude: string[];
}

export const shouldInclude = (type: string, options: ShouldIncludeOptions): boolean => {
  // Determine if 'include' and 'exclude' are effectively set
  const includesEffectivelySet = options.include && options.include.length > 0;
  const excludesEffectivelySet = options.exclude && options.exclude.length > 0;

  // Function to check if any patterns in the array match the type
  const matchesPattern = (patterns: string[], type: string): boolean =>
    patterns.some(pattern => globPattern.test(pattern) ? minimatch(type, pattern) : type === pattern);

  // Check if the type is explicitly included or excluded using minimatch or exact match
  const isIncluded = includesEffectivelySet ? matchesPattern(options.include, type) : true;
  const isExcluded = excludesEffectivelySet ? matchesPattern(options.exclude, type) : false;

  // Apply the logic based on whether includes or excludes are effectively set
  if (includesEffectivelySet && excludesEffectivelySet) {
    return isIncluded && !isExcluded;
  } else if (includesEffectivelySet) {
    return isIncluded;
  } else if (excludesEffectivelySet) {
    return !isExcluded;
  }

  // Default behavior if neither is effectively set
  return true;
};

export const cleanComment = (str: string) => {
  return str.replace(/\*\//g, '*\\/');
};

const ensureOneSpaceEnd = (str: string) => {
  return /[\s\n\t]$/.test(str) ? str : `${str} `;
};

const ensureOneSpace = (str: string) => {
  return /^[\s\n\t]+/.test(str) ? str : ` ${str}`;
};

export const processComment = (comment: string) => {
  if (!comment) return '';

  if (!/[\n]+/.test(comment)) {
    return `*${ensureOneSpaceEnd(ensureOneSpace(cleanComment(comment)))}`;
  }

  let lines = comment.split('\n');
  lines = ['*', ...lines.map(line => cleanComment(line).trim()), ' ']; // Clean and trim each line
  return lines.map((line, i) => {
    if (i === 0) return line;
    if (i === lines.length - 1) return line; // The last line is just a space
    return ` *${ensureOneSpace(line)}`; // Ensure a space after * for clean formatting
  }).join('\n');
};

export const makeComment = (comment: string) => {
  return [{ type: 'CommentBlock', value: ` ${comment} ` }];
};

export const makeCommentLine = (comment: string): t.CommentLine[] => {
  // @ts-ignore
  return [{ type: 'CommentBlock', value: ` ${comment} ` }] as t.CommentLine[];
};