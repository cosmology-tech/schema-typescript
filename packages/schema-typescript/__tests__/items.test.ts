import { generateTypeScript } from '../src';

const schema = {
    "$id": "https://example.com/person.schema.json",
    "$schema": "https://json-schema.org/draft-07/schema#",
    "title": "Person",
    "type": "object",
    "properties": {
        "traces": {
            "type": "array",
            "description": "The origin of the asset, capturing all transitions in form and location.",
            "items": {
                "anyOf": [
                    { "$ref": "#/$defs/dog" },
                    { "$ref": "#/$defs/cat" }
                ]
            },
            "minContains": 1
        },
        "profile": {
            "allOf": [
                { "$ref": "#/$defs/dog" },
                {
                    "type": "object",
                    "properties": {
                        "microchipped": {
                            "type": "boolean"
                        }
                    }
                }
            ]
        },
        "identity": {
            "oneOf": [
                {
                    "type": "object",
                    "properties": {
                        "passportNumber": {
                            "type": "string"
                        }
                    }
                },
                {
                    "type": "object",
                    "properties": {
                        "driverLicense": {
                            "type": "string"
                        }
                    }
                }
            ]
        }
    },
    "$defs": {
        "dog": {
            "type": "object",
            "properties": {
                "name": {
                    "type": "string"
                },
                "breed": {
                    "type": "string"
                },
                "age": {
                    "type": "integer"
                }
            },
            "required": ["name", "breed"]
        },
        "cat": {
            "type": "object",
            "properties": {
                "name": {
                    "type": "string"
                },
                "breed": {
                    "type": "string"
                },
                "age": {
                    "type": "integer"
                }
            },
            "required": ["name", "breed"]
        }
    }
};

it('works', () => {
    expect(generateTypeScript(schema as any)).toMatchSnapshot();
});
