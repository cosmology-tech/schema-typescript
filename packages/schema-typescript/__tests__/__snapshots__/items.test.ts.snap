// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
"export interface Dog {
  name: string;
  breed: string;
  age?: number;
}
export interface Cat {
  name: string;
  breed: string;
  age?: number;
}
export interface Person {
  traces?: (Dog | Cat)[];
  profile?: Dog & {
    microchipped?: boolean;
  };
  identity?: {
    passportNumber?: string;
  } | {
    driverLicense?: string;
  };
}"
`;
